### ISSUE TITLE
[Request] Number formatting filter

### ISSUE THREAD

**Issue by chimericdream** [Shopify/liquid#567]
I was surprised that there isn't a number format filter similar to the date format. I am new to both Ruby and Liquid, but looking through the code it looks like this would not be difficult to add. I plan to give it a try, but with my current schedule I don't know when I can sit down and poke around with it.

I figured I would post here and see if anyone else was missing this and wanted to build it.

**Comment by nickpearson** [comment]
Here's a _very_ simple formatting filter. It depends on Rails, doesn't take any formatting options, and doesn't do input type checking, but it should at least get you on the right track for making custom filters.

```ruby
module CustomFilters
  class ActionViewHelpers
    include Singleton
    include ActionView::Helpers
  end

  def number_with_delimiter(number)
    ActionViewHelpers.instance.number_with_delimiter(number)
  end
end

template = Liquid::Template.parse('{{ 1000 | number_with_delimiter }}')
template.render({}, filters: CustomFilters)
#=> "1,000"
```

**Comment by fw42** [comment]
What exactly do you mean by number formatting? something like `sprintf`?

**Comment by chimericdream** [comment]
@fw42 Basically, yes. As I said, I'm new to Ruby, but I have used JavaScript and PHP extensively and both have number formatting utilities. `sprintf` can do much of the same stuff, but it doesn't have some of the extra functionality like locales.

[JavaScript: `Intl.NumberFormat`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NumberFormat)
[PHP: `number_format`](http://php.net/number_format)

You can get much of the same functionality with `sprintf`, but I was curious if there was something in Ruby itself that was similar to the above (I haven't found anything so far, so my guess is no).

@nickpearson, I am working on a Jekyll site, so I don't know if that will work in my environment. I think it gives me a starting point for making my own Liquid filter, though.

**Comment by felicitus** [comment]
One use case is the following:

I have a number 0.1. I want to multiply it by 1000. The resulting output is 100.0, instead of the expected 100.

**Comment by fw42** [comment]
Hm I would say 100.0 is the expected result of 0.1 * 1000

**Comment by chimericdream** [comment]
@fw42 I agree, but if you wanted to format the number to remove the redundant decimal, you'd need some sort of formatting filter.

This has completely fallen off my plate, and I want to say I worked around it for my needs (been a busy few months), but it'd still be nice to see in a future release.

**Comment by lynxlynxlynx** [comment]
I agree that it would be useful. While the above example could easily be done with rounding, I have a case where I want to display both the decimal and hexadecimal version of an integer. It could be done by manual implementation of the conversion within liquid, but repeating that for the over 1000 uses would be plain silly.

**Comment by thybzi** [comment]
Please make this `sprintf` filter. agree it should be very useful

**Comment by johnteske** [comment]
A Liquid-only solution: https://gist.github.com/johnteske/aab61e8a43ca54dc30ac04888a29cbf1

**Comment by hexerei** [comment]
thanks @johnteske nice gist, which actually inspired me to make a version allowing decimals and conversion from us to european number format (which can be adapted by the defaults), should be close to what @chimericdream was looking for:
https://gist.github.com/hexerei/5bd632b2a179717e219fbe18c5793181

**Comment by carolineschnapp** [comment]
Is it fair to say that when a **float** without a decimal value is **output** without a filter applied we want to see `10` instead of `10.0`?

**Comment by hexerei** [comment]
The gist I supplied reflects that with the decimal='0' parameter, which will truncate all decimal values, and a unified format for all handled values as desired. But the underlying numf.html liquid code could be adapted to allow variant output depending on the existence of decimal values. I do believe though, that a uniform output is a usual case to increase readability.
